test/e2e/framework/pod/wait.go

```diff
@@ -211,13 +218,83 @@ func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns stri
 	}))
 
 	// An error might not be fatal.
-	if err != nil && notReady >= 0 && notReady <= allowedNotReadyPods {
-		framework.Logf("Number of not-ready pods (%d) is below the allowed threshold (%d).", notReady, allowedNotReadyPods)
+	if nOk+len(otherPods) >= minPods && len(otherPods) <= allowedNotReadyPods {
 		return nil
 	}
 	return err
 }
 
+// WaitForPodsRunningReady waits up to timeout for the following conditions:
+//  1. At least minPods Pods in Namespace ns are Running and Ready
+//  2. No Pods in Namespace ns are not either Ready, Succeeded, or Failed with a Controller
+//
+// An error is returned if either of these conditions are not met within the timeout.
+//
+// It has separate behavior from other 'wait for' pods functions in
+// that it requests the list of pods on every iteration. This is useful, for
+// example, in cluster startup, because the number of pods increases while
+// waiting. All pods that are in SUCCESS state are not counted.
+func WaitForPodsRunningReady(ctx context.Context, c clientset.Interface, ns string, minPods int, timeout time.Duration) error {
```

Comment:
This is a new function that is substantially cleaned up and simplified. Most notably, it doesn't even include `allowedNotReadyPods`, as I think that parameter was the source of most confusion and was only used by one caller (the call in `e2e.go`, which is now using the new function above).